### ISSUE TITLE
HLSL: Buffer<> should probably map to a uniform texel buffer

### ISSUE THREAD

**Issue by punkUser** [KhronosGroup/glslang#795]
Currently when using the HLSL frontend a declaration like:
`Buffer<uint>`
... will generate SPIRV that expects a storage texel buffer to be bound. That's somewhat inconsistent with DX where Buffer objects are read-only and support nearly all formats, while the more general RWBuffer's have a much smaller set of format support.

I believe in this case it should be mapping this to a "uniform texel buffer" in SPIRV/Vulkan instead, as this should be the analogous concept (read-only buffer lookups that can go through texture hardware and thus support the broadest set of formats).

**Comment by johnkslang** [comment]
@steve-lunarg does this make sense?

**Comment by steve-lunarg** [comment]
It's certainly possible, though it's really hard to tell from the documentation, which says nearly nothing.

Presently we do (after a fairly recent PR) mark Buffers with the readonly qualifier, but they are still the same underlying type as RWBuffer.  I found some oblique references to type restrictions on RWBuffers in the past, but with some implication that had been relaxed recently.

Also interested to hear if @dankbaker has anything to add.

**Comment by punkUser** [comment]
Just to support my note with an example here:

`
#define THREADS_PER_GROUP 64

Buffer<uint> BufferTest;
StructuredBuffer<uint2> StructuredBufferTest;
ByteAddressBuffer RawBufferTest;

RWBuffer<uint> RWBufferTest;
RWStructuredBuffer<uint2> RWStructuredBufferTest;
RWByteAddressBuffer RWRawBufferTest;

[numthreads(THREADS_PER_GROUP,1,1)]
void main(uint groupID : SV_GroupID, uint threadID : SV_GroupThreadID)
{
	const uint offset = groupID * THREADS_PER_GROUP + threadID;
		
	RWBufferTest[offset] = BufferTest[offset];
	RWStructuredBufferTest[offset] = StructuredBufferTest[offset];
	RWRawBufferTest.Store(offset, RWRawBufferTest.Load(offset));
}
`

Compiles with FXC to produce these bindings:
`
// Resource Bindings:
//
// Name                                 Type  Format         Dim      HLSL Bind  Count
// ------------------------------ ---------- ------- ----------- -------------- ------
// BufferTest                        texture    uint         buf             t0      1
// StructuredBufferTest              texture  struct         r/o             t1      1
// RWBufferTest                          UAV    uint         buf             u0      1
// RWStructuredBufferTest                UAV  struct         r/w             u1      1
// RWRawBufferTest                       UAV    byte         r/w             u2      1
`
Note that all of the RW* things end up as u# bindings while all of the other resources end up as t#'s including the Buffer<uint>. That's because HLSL differentiates these namespaces based on read (SRV) vs. read/write (UAV) rather than buffer vs. texture, etc. Read-only resources typically have better hardware format support and so on, and this is also true in Vulkan (more formats have UNIFORM_TEXEL_BUFFER than STORAGE_TEXEL_BUFFER).

I think the naming of UNIFORM_TEXEL_BUFFER is probably part of the confusion here, but it's still the best analog to typed buffer SRVs that I can find.

**Comment by johnkslang** [comment]
Agreed that `Buffer<type>` should probably map to a SPIR-V `OpTypeImage` of dimension `Buffer`, when `type` can fit in a `float4` and that's what's traditionally expected.

I'm less clear that, if we do this, whether it should be sampled or not sampled.  Let me know what would be normal.

**Comment by punkUser** [comment]
I'm not entirely sure what "sampled" means, but if it means used with a sampler object/filtered, then it should *not* be sampled. While these buffer lookups do have format conversions, they do *not* have filtering, wrapping, or any of that sampler state-related stuff.

**Comment by johnkslang** [comment]
Okay, sounds like we want to map `Buffer<>` to a SPIR-V `OpTypeImage` with `Sampled` set to 2, meaning it will not be used in conjunction with a sampler.

**Comment by punkUser** [comment]
Yeah I'm not an expert on SPIR-V but I believe that would be consistent with our client logic on DX12 for these HLSL types.

**Comment by steve-lunarg** [comment]
> Agreed that Buffer<type> should probably map to a SPIR-V OpTypeImage of dimension Buffer

> Okay, sounds like we want to map Buffer<> to a SPIR-V OpTypeImage with Sampled set to 2, meaning it will not be used in conjunction with a sampler.

Lots of different sets of terminology around from the different APIs, so I'm struggling to get this all straight.  I though the above should be the current behavior - e.g, for a Buffer<float4> we currently produce this SPIR-V type (with a NonWritable decoration):

`    TypeImage 6(float) Buffer nonsampled format:Rgba32f`

I'm happy to make needed changes around how Buffer<> is mapped, though at the moment I'm pretty fuzzy about what the changes should be.

**Comment by johnkslang** [comment]
Okay, that seems different than 

>  Buffers with the readonly qualifier, but they are still the same underlying type as RWBuffer.

from https://github.com/KhronosGroup/glslang/issues/795#issuecomment-288871702, which is what made me think we need to do something, because they are different in DX (https://github.com/KhronosGroup/glslang/issues/795#issuecomment-288890458) and in SPIR-V we just said one needs to map to an image while the other (I thought) maps to a block in the **Uniform** storage class (traditional OpenGL uniform buffer block).

I see the `Buffer<uint> BufferTest` does map to SPIR-V

```
              20:             TypeImage 6(int) Buffer nonsampled format:R32ui
              21:             TypePointer UniformConstant 20
  22(BufferTest):     21(ptr) Variable UniformConstant
```

Are all good already then?  That is, what is the issue?

**Comment by punkUser** [comment]
In our code the validation layer will definitely complain if we don't map these things as a "STORAGE_BUFFER" instead of a "UNIFORM_BUFFER". I'm not sure what that translates to in terms of the SPIR-V (which I find admittedly very hard to parse in these sorts of cases) but right now it is definitely not doing the right thing for us unless there was a very recent change here.

**Comment by johnkslang** [comment]
What layer/API/language is 

> "STORAGE_BUFFER" instead of a "UNIFORM_BUFFER" 

and what is being compared by what validator? Thanks.

**Comment by steve-lunarg** [comment]
> That's because HLSL differentiates these namespaces based on read (SRV) vs. read/write (UAV) 

> unless there was a very recent change here.

Depends on your definition of "very recent" :).  PR #779, IIRC, added NonWritable decorations to the non-RW Buffer forms.  I think that was around 19-Mar, plus or minus a day.

Based on experiments with FXC, it seems that both Buffer and RWBuffer are restricted to no more than a vec4's worth of data (or similar from a user structure, which glslang does not yet implement), whereas structuredbuffers can be templatized on an arbitrary type, even if bigger.  Given the quoted line about about read vs read/write, there's a NonWritable decoration in the SPIR-V which should be set properly post-#779 (AFAIK) for *Buffer, *StructuredBuffer, and *ByteAddressBuffer, but was not set properly pre-#779.

**Comment by punkUser** [comment]
This is the Vulkan validation layer comparing the "layout" you're providing to the SPIR-V when creating a pipeline.

**Comment by punkUser** [comment]
#779 may be new enough that we don't have it in the stuff I'm testing - we're working on getting newer code in but running into various issues with the preprocessor in the latest drop (filed elsewhere). I'll let you know once we get that resolved if it also fixes this issue.

**Comment by johnkslang** [comment]
We must be talking about different things? In Vulkan, storage buffers are not textures, but I understand this request to be to map an HLSL `Buffer<>` to a SPIR-V image.  If, instead you are asking that glslang map HLSL `Buffer<>` to a Vulkan storage buffer, that seems quite contradictory to DX mapping it to a texture.

I think we need to pick one type (there are six in your example), and state the problem in terms of language that's always prefixed with which domain the language comes from (like "SPIR-V image" or "Vulkan storage buffer" or "HLSL Buffer", etc.)

**Comment by punkUser** [comment]
The mappings I referenced above are what we expect and what appear to be similar hardware capabilities between the APIs. We may be getting confused between what "Image" is in Vulkan vs "Texture" in DX, but Buffers in DX are definitely not "textures" in the sense that they use samplers... they just have a DXGI_FORMAT, similar to how "texel buffers" in Vulkan have a VkFormat w/ type conversions.

So indeed we don't expect Buffer<uint> to map to a VkImage in the API code - we expect it to map to a "uniform texel buffer". That seems to be the most direct equivalent Vulkan type that is a buffer, isn't "sampled", but has a VkFormat and conversions.

Hope that makes some amount of sense...

**Comment by johnkslang** [comment]
Okay, picking HLSL `Buffer<>`, my read on all this is:

A Vulkan *uniform texel buffer* is a SPIR-V **OpTypeImage** that has it's *sampled* operand set to 1, which means there is a sampler.  It corresponds to a GLSL `samplerBuffer`, not a GLSL `imageBuffer`.

So, if you are determined to use a Vulkan *uniform texel buffer* for `Buffer<>`, you also need to use it with a Vulkan/SPIR-V sampler.

On the other hand, if you are determined that `Buffer<>` must be a SPIR-V **OpTypeImage** that is not sampled (*sampled* operand is 2), then you need to use a Vulkan *storage texel buffer*, which is what that corresponds to (GLSL `imageBuffer`).

**Comment by johnkslang** [comment]
BTW, the normal "can write to" distinction in the image space correlates with "storage image", and "not storage image" correlates to "sampled", so if that is the important distinction between `Buffer<>` and `RWBuffer<>`, you do want the distinction to be the SPIR-V **OpTypeImage** *sampled* operand being 1 vs. 2 (sampler vs. storage image), correlating with Vulkan "uniform texel buffer" vs. "storage texel buffer".

**Comment by punkUser** [comment]
Hmm, if that's the case then this is a fundamental missing piece of functionality in Vulkan... we need the type conversion, but we don't have a sampler. It's not about what we are "determined to use" per se, it's about getting at the relevant hardware functionality.

STORAGE_TEXEL_BUFFER is what we have to use today (due to the SPIR-V that is generated), but this is a Read/Write type in Vulkan and thus implementations support it with far fewer formats than UNIFORM_TEXEL_BUFFER, which from a hardware standpoint should effectively support all of the formats. i.e. STORAGE_TEXEL_BUFFER is akin to a RWBuffer<> in HLSL, not a Buffer<>.

Thus I still think we need to use a UNIFORM_TEXEL_BUFFER for these read-only, formatted buffers. We're not going to be calling "Sample" on it at all because that doesn't even exist in DX. Are there ".Load" equivalents for these types in SPIR-V that don't use samplers? Perhaps if SPIR-V is only generating those then that would work?

I must admit I'm still confused by your comment though, as when you set up a texel buffer descriptor in Vulkan you give it a "buffer view", but not a sampler, as per the spec.

**Comment by johnkslang** [comment]
@steve-lunarg it is possible that the "writability" discussion conflated "use a NonWritable decoration" with "use a sampled image instead of a storage image".

**Comment by steve-lunarg** [comment]
> A Vulkan uniform texel buffer is a SPIR-V OpTypeImage that has it's sampled operand set to 1,

Also worth noting the above was the behavior pre-#779, and was changed upon request to an unsampled type (which actually makes more sense to me, since there are no sample methods on HLSL Buffer objects).

**Comment by johnkslang** [comment]
I'm suspecting now that `Buffer<>` should map to **OpTypeImage** with a sampler, but that the texels are accessed using **OpImageFetch** which is a non-sampling operation.

**Comment by punkUser** [comment]
For reference, we'll be binding the descriptor to these using the pTexelBufferView parameter here, which does not have any notion of a sampler:
https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#VkWriteDescriptorSet

So whatever it maps to in SPIR-V it should be compatible with that.

**Comment by johnkslang** [comment]
SPIR-V can require a sampled image type, but the fetch operation does not require a sampler.  Further, there is no use of read/write-style decorations on a sampled image.

It seems like #779 should not have been done.  I suspect conflation of "sampled image" with "needs a sampler" which are not quite equivalent if you are just doing fetches and not doing sampling-style operations.  The fetch does not need a sampler, but you actually want to go through the texture-read path to get the conversions you want, which fetch will do.

**Comment by punkUser** [comment]
That's consistent with how I expect the hardware is operating, yes. Support ~all formats because it's going through the texture sampler vs. going through the storage buffers more general memory path with less support.

If that works in terms of what I'm binding from the vulkan side with UNIFORM_TEXEL_BUFFER and the associated VkBufferView (which we're now quite happy with as it's fairly symmetric to DX), that works for us.

**Comment by johnkslang** [comment]
Yes. I reverted #779, to put us in the closer state to this.

The subtlety might be that, even though I think we could make the SPIR-V work without it, a sampler is needed:  It could be the validator (validly) wants you to use a sampler, even though since the image is used only for fetching, the sampler would always be bypassed.  But, I think it looks like the code is set up to want a combined sampler+image to be the buffer, just because no one knew up front that no sampling operations would be done.

That might be an optimization we could consider on top of this, if we actually do know, strictly, from the HLSL type that only fetch operations will be done, then we can skip expecting a combined sampler+image.  For now, the work-around is to include a dummy sampler.

**Comment by dankbaker** [comment]
We treat Buffer and RWBuffers differently.  Buffers = uniform texel buffers, and RWBuffers = Storage Texel buffers. IMHO, this is the correct mapping from HLSL to Vulkan.

**Comment by johnkslang** [comment]
That matches what this converged on, thanks Dan.

**Comment by punkUser** [comment]
When you say "include a dummy sampler", are you just talking from the SPIR-V POV? Because from talking to @dankbaker I don't think they have to do anything like that from an API POV (and indeed there would be no way to do that from the API as far as I can tell...).

**Comment by johnkslang** [comment]
If it works without, that's great, hope we're all discussing the same thing, since the change just went in. The current SPIR-V looks like it wants a combined sampler+image, not just an image.  It could be set up to take just an image.  The SPIR-V I saw in the commit looks like:

```
14:    TypeImage 6 Buffer sampled format:Rgba32f // image type
15:    TypeSampledImage 14                       // combined sampler+image
16:    TypePointer UniformConstant 15            // API object type; points to sampler+image
17: 16 Variable UniformConstant                  // g_tTexbf4, the object from the API
18: 15 Load 17(g_tTexbf4)                        // load the object from the API
30: 14 Image 18                                  // get the image part out of it
31:  7 ImageFetch 30 29                          // fetch from the image, sans sampler
```

The object loaded from the API is not a pure image, it is a sampler+image.

**Comment by punkUser** [comment]
Hmm I'm concerned that won't work then. We don't use combined samplers/images *at all* (and won't be since there's no way to do them in HLSL), nor do we have a "VkImage" for these objects. We really need something that maps to a VkBufferView, as that's how we bind it. VkBufferView's already have a VkFormat in them, so there must be a way to do this... right? Who can connect the dots on what SPIR-V we need for that Vulkan descriptor?

**Comment by johnkslang** [comment]
If the SPIR-V instead was:

```
14:    TypeImage 6 Buffer sampled format:Rgba32f // image type
16:    TypePointer UniformConstant 14            // API object type; points to image
17: 16 Variable UniformConstant                  // g_tTexbf4, the object from the API
18: 15 Load 17(g_tTexbf4)                        // load the object from the API
31:  7 ImageFetch 18 29                          // fetch from the image
```

It might work without a sampler, if validators don't complain there was no sampler, and if we can make the stack do that (I think requires some work at a couple of levels).

**Comment by punkUser** [comment]
Is there someone we can talk to on the runtime/validation/driver side to ensure we've got this all hooked up correctly?

**Comment by dankbaker** [comment]
I think you have to be able to load data from an image without a sampler, that's how .Load(int) works. I would imagine the code sequence is the same for this

**Comment by steve-lunarg** [comment]
> If the SPIR-V instead was:

I'm not sure if this is the whole story, but there's a sampler.combined flag, which if I force it to be false for this case, results the SPIR-V generator not adding the OpTypeSampledImage.  E.g, experimentally in [combine-test](https://github.com/steve-lunarg/glslang/tree/combine-test).  This code:
```
Buffer <float4> g_tTexbf4;

float4 main() : SV_Target0
{
    return g_tTexbf4.Load(0);
}
```

Yields:

```
              11:             TypeImage 6(float) Buffer sampled format:Rgba32f
              12:             TypePointer UniformConstant 11
   13(g_tTexbf4):     12(ptr) Variable UniformConstant
...
              14:          11 Load 13(g_tTexbf4)
              17:    7(fvec4) ImageFetch 14 16

```

I'm a bit uncertain about the AST changes from samplerBuffer to uniformBuffer though.  Too many different sets of terminology for me to keep straight.

**Comment by johnkslang** [comment]
@steve-lunarg yes, that looks correct. So, we already have a way to drive a samplerless sampled image for doing reads through OpImageFetch.  Seems like if we use that, everything will work as desired.

**Comment by Themaister** [comment]
Oddly enough, samplerBuffer and texelFetch in GLSL also generates a magic sampled image type which doesn't make much sense ...

```
#version 450
layout(location = 0) out vec4 FragColor;
layout(binding = 0) uniform samplerBuffer buf;

void main()
{
   FragColor = texelFetch(buf, 0);
}
```

```
               OpCapability Shader
               OpCapability SampledBuffer
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %main "main" %FragColor
               OpExecutionMode %main OriginUpperLeft
               OpSource GLSL 450
               OpName %main "main"
               OpName %FragColor "FragColor"
               OpName %buf "buf"
               OpDecorate %FragColor Location 0
               OpDecorate %buf DescriptorSet 0
               OpDecorate %buf Binding 0
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
      %float = OpTypeFloat 32
    %v4float = OpTypeVector %float 4
%_ptr_Output_v4float = OpTypePointer Output %v4float
  %FragColor = OpVariable %_ptr_Output_v4float Output
         %10 = OpTypeImage %float Buffer 0 0 0 1 Unknown
         %11 = OpTypeSampledImage %10                                     <-- Why?
%_ptr_UniformConstant_11 = OpTypePointer UniformConstant %11
        %buf = OpVariable %_ptr_UniformConstant_11 UniformConstant
        %int = OpTypeInt 32 1
         %16 = OpConstant %int 0
       %main = OpFunction %void None %3
          %5 = OpLabel
         %14 = OpLoad %11 %buf
         %17 = OpImage %10 %14                              <-- sampler2D is casted back to a texture2D ...?
         %18 = OpImageFetch %v4float %17 %16
               OpStore %FragColor %18
               OpReturn
               OpFunctionEnd

```

The spec does do this in the "sample" code for UNIFORM_TEXEL_BUFFER, so I'm not sure if it's possible to *not* use the dummy SampledImage here. It seems more logical to create a variable of the Image type directly instead of SampledImage, but this is a quite significant change which I fear may break stuff.

**Comment by steve-lunarg** [comment]
> So, we already have a way to drive a samplerless sampled image ... Seems like if we use that, everything will work as desired.

I made it a PR #811.  It only impacts Buffer types.

**Comment by johnkslang** [comment]
The specification issue about the example using a sampler when it doesn't make sense is now an internal Khronos issue. It seems the current state of glslang for HLSL is correct though, and the example in the spec. is incorrect. (c7fd73b784c99f97926a230a6bd2ce90e70f9f78 brings in #811 to do that.)

I think the only real pending issues now is whether the full set of types is being correctly handled, given the new confidence in the way `Buffer<...>` is being handled.

**Comment by punkUser** [comment]
Just wanted to update that this seems to work the way that we are expecting now, at least with AMD's driver. The validation layer no longer complains when we bind the associated buffer view as a UNIFORM_TEXEL_BUFFER.

Closing this issue. Thanks!
